client/src/includes/tabs.test.js

```diff
@@ -0,0 +1,96 @@
+
+describe('TabInterface', () => {
+  let tabList;
+  let tab1;
+  let tab2;
+  let panel1;
+  let panel2;
+  let tabInterface;
+
+  beforeEach(() => {
+    document.body.innerHTML = BASIC_TAB_HTML;
+    tabList = document.getElementById('tabs');
+    tab1 = document.getElementById('t1-label');
+    tab2 = document.getElementById('t2-label');
+    panel1 = document.getElementById('t1');
+    panel2 = document.getElementById('t2');
+    tabInterface = new TabInterface(tabList);
+  });
+
+  afterEach(() => {
+    document.getElementsByTagName('html')[0].innerHTML = '';
+    tabInterface = null;
+  });
+
+  it('exists', () => expect(TabInterface).toBeDefined());
+
+  it('instantiates', () => {
+    expect(tabList.wagtailTabInterface).toBeInstanceOf(TabInterface);
+  });
+
+  it('adds accessibility properties', () => {
+    expect(tab1.getAttribute('aria-controls')).toEqual('t1');
+    expect(tab2.getAttribute('aria-controls')).toEqual('t2');
+    expect(tab1.getAttribute('aria-selected')).toEqual('true');
+    expect(tab2.getAttribute('aria-selected')).toEqual('false');
+    expect(panel1.getAttribute('aria-labelledby')).toEqual('t1-label');
+    expect(panel2.getAttribute('aria-labelledby')).toEqual('t2-label');
+    expect(panel1.hasAttribute('hidden')).toEqual(false);
+    expect(panel2.hasAttribute('hidden')).toEqual(true);
+  });
+
+  it('defaults to first tab active', () => {
+    // Ensure tabInterface reports correctly active tab
+    expect(tabInterface.activeTab).toEqual(tab1);
+    // Ensure properties are set correctly on the tabs
+    expect(tab1.classList.contains('active')).toEqual(true);
+    expect(tab1.getAttribute('aria-selected')).toEqual('true');
+    expect(tab2.classList.contains('active')).toEqual(false);
+    expect(tab2.getAttribute('aria-selected')).toEqual('false');
+    // Ensure properties are set correctly on the panels
+    expect(panel1.classList.contains('active')).toEqual(true);
+    expect(panel1.hasAttribute('hidden')).toEqual(false);
+    expect(panel2.classList.contains('active')).toEqual(false);
+    expect(panel2.hasAttribute('hidden')).toEqual(true);
+  });
+
+  it('lets you change tabs on click', () => {
+    tab2.click();
+    // Ensure tabInterface reports correctly active tab
+    expect(tabInterface.activeTab).toEqual(tab2);
+    // Ensure properties are changed correctly on the tabs
+    expect(tab1.classList.contains('active')).toEqual(false);
+    expect(tab1.getAttribute('aria-selected')).toEqual('false');
+    expect(tab2.classList.contains('active')).toEqual(true);
+    expect(tab2.getAttribute('aria-selected')).toEqual('true');
+    // Ensure properties are changed correctly on the panels
+    expect(panel1.classList.contains('active')).toEqual(false);
+    expect(panel1.hasAttribute('hidden')).toEqual(true);
+    expect(panel2.classList.contains('active')).toEqual(true);
+    expect(panel2.hasAttribute('hidden')).toEqual(false);
+  });
+
+  it('changes to next tab on right arrow press', () => {
```

Comment:
For some reason this test fails not because the keydown doesn't work, but because at this point `tab2` is already selected to start with, despite the fact that I am trying to reset the DOM in `afterEach` every time. I can't figure out why, but somehow the state of which tab is selected seems to be persisting between tests! Any guidance would be welcome.

Reply:
Would this be because your code resets then HTML and re-initialises the tab behavior, but doesn’t reset `window.location.hash`?